keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -893,6 +894,48 @@ of the IP family of the service. A single service which is dual-stack may use
 the same NodePort on both families.  `kube-proxy` will ensure that traffic is
 routed according to the families assigned for services.
 
+
+##### Spec.loadBalancerIP
+
+The `spec.loadBalancerIP` is a string and only allow one load-balancer
+address to be specified. A new multi value `spec.loadBalancerIPs` is added;
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: LoadBalancer
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+  ipFamilyPolicy: PreferDualStack
+  loadBalancerIPs:
```

Comment:
can i ask you to layout transformation rules singular->plural. Specifically
1. What happens in core version?
2. How are fields will be validated (and mutated)?
3. Can a user mutate LB type service (to ClusterIP or other types)? if then what happens to field values (who resets what)? and how would a cloud provider know about the old reset values (if applicable).

Reply:
1. Existing services are single-stack. loadBalancerIP -> loadBalancerIPs[0] (if set)
2. What is "core version"?
3. That's the question. I wanted the same validation (or lack of) as for loadBalancerIP; only address validation. The other alternative is 1-2 addresses, if 2 then one of each family.
4. This question must exist in exactly the same way for the existing loadBalancerIP. I assume the same rules for loadBalancerIPs. I suspect the field is left as-is.

Reply:
Probably a check that a user does not specify both loadBalancerIP and loadBalancerIPs in the service manifest should be added. At least not contradicting loadBalancerIP != loadBalancerIPs[0].

Reply:
So we THINK we have the protocol for pluralizing a field pinned down.  This could be a good test of it.

1) add the plural field to the type, leave the old field
1a) document that when specified, plural[0] must be the same as singular
2) add logic to "normalize" the two fields in the context of a create vs update
2a) see NormalizeClusterIPs() as an example
3) call this normalization from defaultOnReadService()
4) call this normalization from PrepareForCreate() and PrepareForUpdate()
5) add validation that singular and plural[0] are in sync

PTAL at https://github.com/kubernetes/kubernetes/pull/95894 which makes default-on-read a thing for ClusterIPs.